### DISCUSSION TITLE
Easier way to include additional built-in plugins in a custom Conduit build

### DISCUSSION THREAD

**Discussion by hariso** [ConduitIO/conduit#1714] [Ideas]
**The problem**
Currently, to include additional built-in plugins into a custom Conduit build, the source code needs to be changed. That makes it more complicated to pull in new Conduit changes (revert changes to the list of built-in connector, pull new Conduit changes, add new built-in connectors again, build Conduit).

**The 'why'**

Built-in plugins are faster than standalone plugins, so there's a strong motivation to include a plugin in Conduit's binary. I'm also assuming that the connectors used for a specific Conduit instance are not going to change that frequently. In those cases, where the connectors used are all/most of the time the same, it makes sense to have a custom build with those connectors as built-in plugins.

**An example**

The list of built-in plugins can be made "external", i.e. in a separate file and then included during the build process, for example:
```
/path/to/plugins.go

import (
   _ github.com/conduitio-labs/conduit-connector-foo
)
```

Then, something like `make build-with-plugins /path/to/plugins.go` would build Conduit with conduit-connector-foo as a built-in plugin.

For now, I think this discussion can focus on the usefulness of such a feature. If we agree that having it is useful, we can discuss the implementation.

**Comment by lovromazgon** [comment]
I don't entirely understand the problem. Do you mean that it's hard to exclude built-in plugins? Or are we trying to make it easier to include additional plugins? Or is it updating built-in plugins themselves? Essentially, I'm not sure what "pull in new Conduit changes" means and what we're improving.

**Reply by hariso** [reply]
So, as mentioned in the introductory paragraph, it's about including additional built-in plugins. "Pulling new Conduit changes" means pulling new changes in Conduit's source code (before making another custom build).

**Reply by lovromazgon** [reply]
Ok, so if I understand correctly, that can be done using the [`ConnectorPlugins`](https://github.com/ConduitIO/conduit/blob/1ad15ae2e52d7852777b9228f3b8c39e1e5c368d/pkg/conduit/config.go#L88) field, something like this:

```go
package main

import (
	"github.com/conduitio/conduit/pkg/conduit"
	"github.com/conduitio-labs/conduit-connector-foo"
)

func main() {
	// get the default config including all builtin plugins
	cfg := conduit.DefaultConfig()
	
	// add additional builtin plugins here
	cfg.ConnectorPlugins["foo"] = foo.Connector
	
	conduit.Serve(cfg)
}
```

Pulling new changes can be done by updating the conduit dependency, no need to change the source code.

**Reply by hariso** [reply]
For the embedded mode, yes. What about a custom build though?

**Reply by lovromazgon** [reply]
Define custom build please, I thought that's what I did above 😄 The default Conduit [entrypoint](https://github.com/ConduitIO/conduit/blob/1ad15ae2e52d7852777b9228f3b8c39e1e5c368d/cmd/conduit/main.go#L20) doesn't do anything else, so with the above code you get exactly the same behavior as Conduit + one additional builtin connector plugin.

**Reply by hariso** [reply]
Oh, I see now, I read it too quickly.:) I was looking for flag parsing and similar, I missed that it's in Serve. This looks good, I think I'll something in the docs.

**Reply by lovromazgon** [reply]
We could definitely add a doc entry for how to build a custom conduit instance with additional builtin plugins.
